### PR TITLE
ansible modules for IPMI, SOL, Server Power and LED locator

### PR DESCRIPTION
**Pull request by sandinak**

### CODE REVIEW

**Comment by sandinak** [comment]
This has:
- cisco_imc_ipmi - enable and disable ipmi access
- cisco_imc_sol - enable and disable sol on the cimc
- cisco_imc_server - change power state and indicator_led

File: library/cisco_imc_ipmi.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  cisco_imc_ipmi:
+    priv: "admin"
+    state: "present"
+    ip: "192.168.1.1"
+    username: "admin"
+    password: "password"
+'''
+
+
+def login(module):
```
Comment (vvb) [inline library/cisco_imc_ipmi.py]: `login` and `logout` are implemented in common code. I see that you are using the below code later. so you can safely remove `login` and `logout` from this file.

```
from ansible.module_utils.cisco_imc import ImcConnection
conn = ImcConnection(module)                                 
server = conn.login()                                                 
conn.logout()
```

File: library/cisco_imc_ipmi.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                           secure=ansible["secure"],
+                           proxy=ansible["proxy"])
+        server.login()
+    except Exception as e:
+        results["msg"] = str(e)
+        module.fail_json(**results)
+    return server
+
+
+def logout(module, imc_server):
```
Comment (vvb) [inline library/cisco_imc_ipmi.py]: same as above

File: library/cisco_imc_ipmi.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    if imc_server:
+        imc_server.logout()
+        return True
+    return False
+
+
+def setup_ipmi(server, module):
+    from imcsdk.apis.admin.ipmi import ipmi_disable
+    from imcsdk.apis.admin.ipmi import ipmi_enable
+    from imcsdk.apis.admin.ipmi import is_ipmi_enabled
```
Comment (vvb) [inline library/cisco_imc_ipmi.py:52]: @waghswapnil  this is more of a comment on API layer provided by imcsdk - `is_ipmi_enabled` -> `ipmi_is_enabled`

File: library/cisco_imc_server.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    indicator_led: on
+    poweroff_seconds: 300
+    ip: "192.168.1.1"
+    username: "admin"
+    password: "password"
+
+'''
+
+
+def login(module):
```
Comment (vvb) [inline library/cisco_imc_server.py]: same as before

File: library/cisco_imc_server.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+                           secure=ansible["secure"],
+                           proxy=ansible["proxy"])
+        server.login()
+    except Exception as e:
+        results["msg"] = str(e)
+        module.fail_json(**results)
+    return server
+
+
+def logout(module, imc_server):
```
Comment (vvb) [inline library/cisco_imc_server.py]: same as before

File: library/cisco_imc_server.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    return results, err
+
+
+def main():
+    from ansible.module_utils.cisco_imc import ImcConnection
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_id=dict(required=False, type='int', default=1),
+            chassis_id=dict(required=False, type='int', default=1),
+            state=dict(required=False, type='str',
```
Comment (vvb) [inline library/cisco_imc_server.py:156]: should this be `status`?

File: library/cisco_imc_server.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        locator_led_off(server, 
+                        server_id=server_id, 
+                        chassis_id=chassis_id)
+    return True
+
+    
+
+def setup_server(server, module):
+    ansible = module.params
+    status, locator_led = ansible["status"], ansible["locator_led"]
```
Comment (vvb) [inline library/cisco_imc_server.py:121]: `locator_led` does not seem to exist in ansible parameters.

File: library/cisco_imc_server.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+    
+    if status is not None:
+        power_changed = setup_server_power(server, module, 
+                               status, timeout, interval)
```
Comment (vvb) [inline library/cisco_imc_server.py]: where are `timeout` and `interval` coming from?

File: library/cisco_imc_server.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        
+    if locator_led is not None:
```
Comment (vvb) [inline library/cisco_imc_server.py]: `check_mode` handling is not done here.

File: library/cisco_imc_sol.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    speed: "9600"
+    ssh_port: 22
+    state: "present"
+    ip: "192.168.1.1"
+    username: "admin"
+    password: "password"
+'''
+
+
+def login(module):
```
Comment (vvb) [inline library/cisco_imc_sol.py]: `login` method is not required

File: library/cisco_imc_sol.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                           secure=ansible["secure"],
+                           proxy=ansible["proxy"])
+        server.login()
+    except Exception as e:
+        results["msg"] = str(e)
+        module.fail_json(**results)
+    return server
+
+
+def logout(module, imc_server):
```
Comment (vvb) [inline library/cisco_imc_sol.py]: `logout` method is not required

**Review by vvb — CHANGES_REQUESTED** [review]
@sandinak thank you for writing up these modules. You have got most of them correct with a minor few nits. I have provided comments where due.

File: library/cisco_imc_server.py (hunk 6 above)
Comment (sandinak) [reply library/cisco_imc_server.py:156]: I don't' believe so.  status implies "position of affairs" .. state implies "particular condition at a time". Most of the ansible modules use 'state' as the moniker: file, service, apt, yum. The only one that uses 'status' AFAIK is the return value of 'stat'.

File: library/cisco_imc_ipmi.py (hunk 3 above)
Comment (sandinak) [reply library/cisco_imc_ipmi.py:52]: Yup .. if it's fixed we should update this as well.

File: library/cisco_imc_server.py (hunk 7 above)
Comment (sandinak) [reply library/cisco_imc_server.py:121]: fixed

File: library/cisco_imc_server.py (hunk 8 above)
Comment (sandinak) [reply library/cisco_imc_server.py]: fixed and updated docs.

File: library/cisco_imc_ipmi.py (hunk 1 above)
Comment (sandinak) [reply library/cisco_imc_ipmi.py]: this was copied directly from cisco_imc_ntp. Probably should be cleaned up as well

File: library/cisco_imc_server.py (hunk 6 above)
Comment (vvb) [reply library/cisco_imc_server.py:156]: well, the original comment was given because there were duplication of `state` variables.. now it makes sense.

**Pull request merged**